modules/apps/batch-engine/batch-engine-service/src/main/java/com/liferay/batch/engine/internal/BatchEngineTaskExecutorImpl.java

```diff
@@ -121,10 +121,9 @@ private void _execute(BatchEngineTask batchEngineTask) throws Throwable {
 				if (items.size() < batchEngineTask.getBatchSize()) {
 					items.add(item);
 				}
-				else {
-					_commitItems(
-						batchEngineTaskItemWriter, items,
-						batchEngineTaskOperation);
+
```

Comment:
Not sure what do you mean here, isn't that "if (!items.isEmpty())" right after this while loop is catching up the tail?

Reply:
@shuyangzhou this problem is, that last statement only saves the last batch of unsaved items, but for all batches you will loose one record per batch group

Reply:
@shuyangzhou , sorry, I see that resource instance is still created in addService method. I was misguided with the return type of addingService method. Everything is ok, and it was my mistake to not to release it in my original PR.